### PR TITLE
041 Authentication Implementation

### PR DESCRIPTION
**Pull request by eoinmcbennett**

### CODE REVIEW

**Review by 40175096 — APPROVED** [review]

File: src/test/java/org/kainos/ea/team2/integration/JobIntegrationTest.java [hunk 1]
```diff
@@ -27,20 +28,33 @@ public class JobIntegrationTest {
...
      * Verify that the getJobs method returns a list of jobs from the database.
      */
     @Test
     void getJobs_shouldReturnListOfJobs() {
+        String jwt = getJWT();
 
         // list of employees, add each employee returned from the db
         List<Job> response = APP.client().target("http://localhost:8080/api/job-roles")
-                .request().get(List.class);
+                .request().header("Authorization", "Bearer " + jwt).get(List.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/team2/integration/JobIntegrationTest.java:109]: can you add integration tests for each endpoint without this header and verify unauthorized returned?

File: src/main/java/org/kainos/ea/team2/cli/Authorise.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(value = RetentionPolicy.RUNTIME)
+@NameBinding
+public @interface Authorise {
+    /**
+     * Stores the passed in role for the annotation.
+     * @return the lowest role that can access the resource
+     */
+    UserRole value() default UserRole.User;
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/team2/cli/Authorise.java]: should we be defaulting this? wouldn't that provide access to endpoints they shouldn't be able to access?

File: src/main/java/org/kainos/ea/team2/resources/JobController.java [hunk 3]
```diff
@@ -31,6 +36,7 @@ public class JobController {
     @GET
     @Path("/job-roles")
     @Produces(MediaType.APPLICATION_JSON)
+    @Authorise(UserRole.User)
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/team2/resources/JobController.java]: admin should also have access to this endpoint

File: src/main/java/org/kainos/ea/team2/resources/JobController.java [hunk 4]
```diff
@@ -54,7 +60,9 @@ public Response getJobs() {
     @GET
     @Path("/job-specification/{id}")
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getJobSpec(@PathParam("id") final int id) {
+    @Authorise(UserRole.User)
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/team2/resources/JobController.java]: admin should also have access to this endpoint

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]
if you had to add a new role I think it would break all of this auth but we can call that tech debt for now :)

**Pull request merged**